### DISCUSSION TITLE
How can one enforce a left boundary condition in _make_derivative(some_grid, method="forward")?

### DISCUSSION THREAD

**Discussion by HannoSpreeuw** [zwicker-group/py-pde#312] [Q&A] [answered]
I have been running this code, with Depths a `CartesianGrid` and cCa a `ScalarField` which uses the `Depths` grid
```
self.forward_diff = _make_derivative(Depths, method="forward")
cCa_grad_forw = cCa.copy()
cCa_grad_forw.data[:] = 0
cCa.set_ghost_cells(self.bc_cCa)
self.forward_diff(cCa._data_full, out = cCa_grad_forw.data)
```

When looking at `cCa_grad_forw` or rather on how the pde integrates using this code, it seems the left bc for `cCa`, from `self.bc_cCa`, is not imposed, probably due to the way `_make_derivative` with `method="forward`" is 
implemented. 
Looking at the [source code](https://py-pde.readthedocs.io/en/latest/_modules/pde/grids/operators/cartesian.html#):

```
elif method == "forward":
    out[i - 1] = (arr[i + 1] - arr[i]) / dx
```
It seems the 'left' side of the data, which would have to come from `arr[i-1]` and taken from a ghost cell, is never read when forward differencing is applied?

How to fix this?

**Comment by david-zwicker** [comment, marked as answer]
Yes, this is expected behavior. If you only use a single derivative, imposing a boundary conditions on one side is sufficient.

**Reply by HannoSpreeuw** [reply]
Thanks, but now I get solutions which do not respect the boundary condition.

I do, however, want to use a forward difference, to gain better accuracy.
I am trying to account of advection, in this case a flow from higher depths to lower depths, so I would like to make use of upstream values.

**Reply by HannoSpreeuw** [reply]
I could redefine all derivatives.
```
                if method == "central":
                    out[i - 1] = (arr[i] - arr[i - 2]) / (2 * dx)
                elif method == "forward":
                    out[i - 1] = (arr[i] - arr[i-1]) / dx
                elif method == "backward":
                    out[i - 1] = (arr[i-1] - arr[i - 2]) / dx
```

**Reply by david-zwicker** [reply]
I guess I understand what you're trying to do, but I don't think the implementation you're suggesting is correct. This is because the index into the `out` and `arr` fields are not directly comparable. The `out` array has `N` entries corresponding to the `N` support points in this 1D case. In contrast, the `arr` field has `N+2` entries, because it accounts for the virtual points (or ghost points) outside either boundary. Consequently, the entry `arr[i]` co-localizes with point `out[i - 1]` and the original definitions are the proper forward, backward, and central differences. I agree that this can be confusing and a better comment would go a long way.
